server/runtime/step_executor.go

```
@@ -355,6 +312,147 @@ func (s *defaultStepExecutor) RunStepLoopWithOptions(ctx context.Context, stepID
+		return preparedCompletedResponse{}, err
+	}
+	hostedToolExecutions := hostedToolExecutionsFromOutputItems(resp.OutputItems, tools.DefinitionsFor(shape.EnabledTools))
+	phaseTurn := s.phase.Apply(ctx, resp, resp.Assistant, localToolCalls, hostedToolExecutions)
+	assistantMsg := phaseTurn.Assistant
+	localToolCalls = phaseTurn.LocalToolCalls
+	hostedToolExecutions = phaseTurn.HostedToolExecutions
+	executedToolCall := len(localToolCalls) > 0 || len(hostedToolExecutions) > 0
+	noopFinalAnswer := isNoopFinalAnswer(assistantMsg)
+
+	if preflightErr := workflowPreflightError(e.workflowRunActive(), localToolCalls, hostedToolExecutions); preflightErr != nil {
+		return preparedCompletedResponse{
```

golangci-lint (2.12.2)

[error] error is not nil (line 341) but it returns nil

(nilerr)